### PR TITLE
enable reload auditd rules

### PR DESCRIPTION
**Pull request by BrookHF**
Fulfilled an open issue https://github.com/elastic/beats/issues/8556
Reload `/etc/auditbeat/audit.rules.d/*` files same way as `/etc/auditbeat/modules.d/*yml`

@vjsamuel @andrewkroh Would you help review and merge the PR?

### CODE REVIEW

File: auditbeat/module/auditd/audit_linux.go
Comment (andrewkroh) [inline auditbeat/module/auditd/audit_linux.go]: I think it would be useful for debugging purposes to also log when it exits.

```
ms.log.Info("audit_rule_files reload watcher started")
defer ms.log.Info("audit_rule_files reload watcher stopped")
```

File: auditbeat/module/auditd/audit_linux.go
Comment (andrewkroh) [inline auditbeat/module/auditd/audit_linux.go]: Should the reloader be stopping on error? I imaging that when new errors I deployed failed to load, I will deploy more changes to replace the bad ones.

File: auditbeat/module/auditd/audit_linux.go
Comment (andrewkroh) [inline auditbeat/module/auditd/audit_linux.go]: This is the wrong format for `Errorw`. Down below the format is correct.

**Comment by andrewkroh** [comment]
jenkins, test this

**Comment by ndubss** [comment]
Looks like this is getting close! Thanks to all that have contributed.

File: auditbeat/module/auditd/config.go
Comment (adriansr) [inline auditbeat/module/auditd/config.go]: According to this logic, the following config will result in reload being enabled:
```yaml
reload:
  enabled: false
  period: 1m
```

This is misleading. I understand that what you want is for this to be implicitly enabled:
```yaml
reload:
  period: 1m
```

For that you can:
1. Remove the enabled flag and let the value of `period` to control if the feature is enabled or not.
2. Set enabled=true by default in `defaultConfig` above, and change the `||` in this logic to a `&&`.
3. Make Enabled a *bool instead of a bool, then you can:
`return r != nil && r.Period!=0 && (r.Enabled == nil || *r.Enabled == true)`

Comment (BrookHF) [reply auditbeat/module/auditd/config.go]: Sorry for the late reply. After having the `defaultConfig`, the logic looks much cleaner for me.

**Comment by danmx** [comment]
you could reload rules when they change using `inotify` (module [fsnotify](https://github.com/fsnotify/fsnotify)) not scan for changes

**Comment by ndubss** [comment]
This has been working for us. We are now able to change configs of rules and yaml dynamically without killing pods.

**Comment by vjsamuel** [comment]
@danmx one issue is that inotify doesnt agree with configmaps on kubernetes. would a refresh interval be fine to begin with?

**Comment by danmx** [comment]
@vjsamuel true then maybe periodical read is the way to go

**Comment by vjsamuel** [comment]
@danmx CI seems to be passing now. can we move forward with this?

**Comment by danmx** [comment]
@vjsamuel I'm not the maintainer just interested in the functionality 😄  so I'd recommend poking @andrewkroh

**Comment by andrewkroh** [comment]
@vjsamuel @BrookHF The changes I [requested](https://github.com/elastic/beats/pull/10835#pullrequestreview-210996195) still need to be addressed. Most importantly this needs to have tests that validate it works as expected.

I see that this is monitoring the file system and refreshing the in memory representation of the rules, but I think something is also needed to apply the updated rules to the kernel, no?

**Comment by vjsamuel** [comment]
@andrewkroh apologies, i had assumed all comments have been taken care of. @BrookHF can you please do the needful.

**Comment by BrookHF** [comment]
Thanks @vjsamuel for bringing this up, maybe there is a misunderstanding. And thanks for reviewing this PR @andrewkroh. I thought about unit testing this change very hard for a period of time. I read the code base related to the config reload. I am blocked in the process of creating the unit test. 

To change the rules on the kernel audit module, it will create a new libaudit client instant to do the rule change. So there is not shared resource on the beats side. The thing we have to test is maybe there are more than one libaudit clients contacting to kernel audit module. And I am not sure how to do it.

Can anyone shed some light on this problem? I will think about it more in my spare time.

**Comment by BrookHF** [comment]
:+1 will come back for the unit test

**Comment by BrookHF** [comment]
:+1

**Comment by BrookHF** [comment]
- [x] Add unit tests. Run them with the race detector enabled.
- [x] Fix the Hound CI warnings.
- [x] Address the review items.
- [x] Update the docs to include the new option.
- [x] Update the reference config files to include the option.
- [x] Add an entry to the CHANGELOG.next.asciidoc file.
- [ ] I'll take another look and I'll do some hands on testing.
- [ ] Merge

**Comment by BrookHF** [comment]
After running the test with the race condition detector, there is a data race detected. From my understanding, it is because the reloader goroutine is trying to change `config.auditrules` while the consumer goroutine is calling `func buildMetricbeatEvent(msgs []*auparse.AuditMessage, config Config) mb.Event`. As golang is passed by value, and it has to make a copy of the entire config struct calling the function, which require it to read the field `config.auditrules`.

By making the argument `func buildMetricbeatEvent(msgs []*auparse.AuditMessage, config Config) mb.Event` to use pointer of the config struct `func buildMetricbeatEvent(msgs []*auparse.AuditMessage, *config Config) mb.Event`. It solved the data race.

I think it's better for performance to avoid memory copy by passing pointer instead of struct. Let me know if there is any feedback. @andrewkroh 

```
root@fb9e4704e4f2:/go/src/github.com/elastic/beats/auditbeat# go test -run TestRuleReload github.com/elastic/beats/v7/auditbeat/module/auditd -audit
ok      github.com/elastic/beats/v7/auditbeat/module/auditd     18.777s
root@fb9e4704e4f2:/go/src/github.com/elastic/beats/auditbeat# go test -race -run TestRuleReload github.com/elastic/beats/v7/auditbeat/module/auditd -audit
==================
WARNING: DATA RACE
Write at 0x00c00000c700 by goroutine 42:
  github.com/elastic/beats/v7/auditbeat/module/auditd.(*Config).reloadRules()
      /go/src/github.com/elastic/beats/auditbeat/module/auditd/config_linux.go:133 +0x8b1
  github.com/elastic/beats/v7/auditbeat/module/auditd.(*reloader).Run()
      /go/src/github.com/elastic/beats/auditbeat/module/auditd/audit_linux.go:283 +0x8dc

Previous read at 0x00c00000c700 by goroutine 38:
  github.com/elastic/beats/v7/auditbeat/module/auditd.(*MetricSet).Run.func2()
      /go/src/github.com/elastic/beats/auditbeat/module/auditd/audit_linux.go:207 +0xe4

Goroutine 42 (running) created at:
  github.com/elastic/beats/v7/auditbeat/module/auditd.(*MetricSet).Run()
      /go/src/github.com/elastic/beats/auditbeat/module/auditd/audit_linux.go:216 +0x564

Goroutine 38 (running) created at:
  github.com/elastic/beats/v7/auditbeat/module/auditd.(*MetricSet).Run()
      /go/src/github.com/elastic/beats/auditbeat/module/auditd/audit_linux.go:200 +0x3b0
```

**Comment by BrookHF** [comment]
:+1

**Comment by ndubss** [comment]
This requires manual review. @andrewkroh please take a look if you can.

**Comment by andrewkroh** [comment]
jenkins, run tests

**Comment by BrookHF** [comment]
👍

**Comment by vjsamuel** [comment]
@BrookHF can you please rebase?

**Comment by BrookHF** [comment]
@vjsamuel I rebased code with the latest master branch. Let me know if there is anything needed from me to help merge this long pending PR.

**Comment by andrewstucki** [comment]
@BrookHF I'll take a look at this PR soon and hopefully we can work out the next steps, and potentially merge, this week.

**Comment by BrookHF** [comment]
@andrewstucki Thanks for taking time to have a look at this PR. Let me know if you need anything from me.

File: auditbeat/module/auditd/config_linux.go
Comment (andrewstucki) [inline auditbeat/module/auditd/config_linux.go:95]: We should probably keep this as `false` by default.

File: CHANGELOG.next.asciidoc
Comment (andrewstucki) [inline CHANGELOG.next.asciidoc:588]: Looks like a stray addition in the changelog

**Review by andrewstucki — CHANGES_REQUESTED** [review]
Hi @BrookHF sorry for the delay. Overall this looks like a really good addition. Main comment is that we should probably keep this behavior defaulted to off and let users opt in. Pending that change, the changelog clean-up and any merge conflict resolutions, we can get this merged.

File: auditbeat/docs/modules/auditd.asciidoc
Comment (andrewstucki) [inline auditbeat/docs/modules/auditd.asciidoc:213]: This looks like it's in the middle of the `backpressure_strategy` documentation. Think it needs to go below the `The possible values are` block.

**Comment by BrookHF** [comment]
@andrewstucki Thanks for the suggestions. I made the changes according to it.

File: auditbeat/docs/modules/auditd.asciidoc
Comment (andrewstucki) [inline auditbeat/docs/modules/auditd.asciidoc:140]: can you update this to correspond to the new default?

File: auditbeat/auditbeat.reference.yml
Comment (andrewstucki) [inline auditbeat/auditbeat.reference.yml:42]: same here

File: auditbeat/module/auditd/audit_linux_test.go
Comment (andrewstucki) [inline auditbeat/module/auditd/audit_linux_test.go:329]: pretty sure the lack of a newline here is going to result in a linter error, try making sure to run `mage fmt` in the `auditbeat` directory.

**Review by andrewstucki — COMMENTED** [review]
Almost there, looks like some issues with merge conflicts and a couple of other things to clean up, and then we can get this merged. Thanks again!

File: auditbeat/module/auditd/audit_linux_test.go
Comment (BrookHF) [reply auditbeat/module/auditd/audit_linux_test.go:328]: Thanks spotting this. I ran mage fmt in the auditbeat directory, but did not get any prompt or correction about it. Corrected it anyway.

```
 ~/go/src/github.com/elastic/beats/auditbeat                                                                                                                [13:23:56] 
> $ mage fmt                                                                                                                                           [±feature-reload-auditd-rule-file ✓]
>> fmt - go-licenser: Adding missing headers
>> fmt - goimports: Formatting Go code
>> fmt - autopep8: Formatting Python code
```

File: auditbeat/docs/modules/auditd.asciidoc
Comment (BrookHF) [reply auditbeat/docs/modules/auditd.asciidoc]: 👍

File: auditbeat/auditbeat.reference.yml
Comment (BrookHF) [reply auditbeat/auditbeat.reference.yml]: 👍

**Comment by BrookHF** [comment]
Rebased to latest upstream to resolve the merge conflict.

**Comment by BrookHF** [comment]
Rebased to latest upstream to resolve the merge conflict again.
@andrewkroh @andrewstucki Would you take a look?

**Comment by andrewstucki** [comment]
jenkins run tests

File: CHANGELOG.next.asciidoc
Comment (andrewstucki) [inline CHANGELOG.next.asciidoc:615]: You'll want to remove the stray changelog entries.

File: auditbeat/auditbeat.reference.yml
Comment (andrewstucki) [inline auditbeat/auditbeat.reference.yml:43]: The tests are failing because this file is actually generated, the proper file to add these additions to is `auditbeat/module/auditd/_meta/config.yml.tmpl`

File: auditbeat/docs/modules/auditd.asciidoc
Comment (andrewstucki) [inline auditbeat/docs/modules/auditd.asciidoc:235]: Same as above, these changes should go in: `auditbeat/module/auditd/_meta/docs.asciidoc`. In addition, after both this and the above change are made, you'll need to run `mage update` in the auditbeat directory

Comment (andrewstucki) [inline auditbeat/docs/modules/auditd.asciidoc:235]:
```suggestion
*`reload`*:: `enabled` Causes auditbeat to reload auditd rule files at a fixed interval if any changes to the rule files are detected.
`period`, Specifies the interval after which auditbeat will check and reload rule files.
```

**Review by andrewstucki — CHANGES_REQUESTED** [review]
@BrookHF Sorry, just took a look at the test run, and it looks like I didn't catch before that the documentation and reference file changes were made in the wrong places. It's minor, but you'll need to change the files specified in my comments rather than the generated files.

**Comment by BrookHF** [comment]
@andrewstucki Thanks spotting this. Did the changes accordingly.

**Comment by andrewstucki** [comment]
jenkins run tests

**Comment by ruflin** [comment]
@andrewkroh What is the status of this PR?

**Comment by noway1979** [comment]
Really appreciating this PR and would love to see it re-enacted because it would solve a major headache for us. 
I tried with dynamically reloading auditbeat configuration fragments (https://www.elastic.co/guide/en/beats/auditbeat/current/auditbeat-configuration-reloading.html) but the audit rules coming from added fragments do not seem to be applied cumulatively. Therefore, happy for any possibility how audit rules can be dynamically reloaded without restarting auditbeat.

**Pull request closed without merging**
